app/com/arpnetworking/metrics/portal/scheduling/JobExecutorActor.java

```diff
@@ -96,42 +100,108 @@ public void preStart() throws Exception {
+        }
+        return Optional.empty();
+    }
+
+    private void scheduleExtraTickIfNecessary(final Instant wakeUpBy) {
+        timeUntilExtraTick(wakeUpBy).ifPresent(timeRemaining -> {
+            timers().startSingleTimer(EXTRA_TICK_TIMER_NAME, Tick.INSTANCE, timeRemaining);
+        });
+    }
+
+    private void killSelf() {
+        // TODO(spencerpearson): Per https://doc.akka.io/docs/akka/2.5.4/java/cluster-sharding.html#remembering-entities ,
+        //   might we need to send a Passivate message to our parent?
+        getSelf().tell(PoisonPill.getInstance(), getSelf());
+    }
+
+    /**
+     * Initializes the actor with the given JobRef (if uninitialized), or ensure the the given ref equals the one already initialized with.
+     *
+     * @param ref The JobRef to initialize with.
+     * @return The {@link CachedJob} the actor is initialized to.
+     *   (Guaranteed to equal {@code _cachedJob.get()}; this return value is purely for the typechecker's sake.)
+     * @throws IllegalStateException If the actor was already initialized with a different JobRef.
+     */
+    private CachedJob<T> initializeOrEnsureRefMatch(final JobRef<T> ref) throws IllegalStateException {
+        if (!_cachedJob.isPresent()) {
+            LOGGER.info()
+                    .setMessage("initializing")
+                    .addData("ref", ref)
+                    .log();
+            _cachedJob = Optional.of(new CachedJob<>(_injector, ref));
+        }
+
+        final JobRef<T> oldRef = _cachedJob.get().getRef();
+        if (!oldRef.equals(ref)) {
+            LOGGER.error()
+                    .setMessage("JobRef in received message does not match cached JobRef")
+                    .addData("cached", oldRef)
+                    .addData("new", ref)
+                    .log();
+            killSelf();
+            throw new IllegalStateException(String.format("got JobRef %s, but already initialized with %s", ref, oldRef));
         }
+
+        return _cachedJob.get();
     }
 
-    private void scheduleNextTick(final Instant wakeUpBy) {
-        timers().startSingleTimer("TICK", Tick.INSTANCE, timeUntilNextTick(wakeUpBy));
+
+    private JobRef<T> unsafeJobRefCast(@SuppressWarnings("rawtypes") final JobRef ref) {
+        // THIS MAKES ME SO SAD. But there's simply no way to plumb the type information through Akka.
+        @SuppressWarnings("unchecked")
+        final JobRef<T> typedRef = ref;
+        return typedRef;
     }
 
-    private void executeAndScheduleNextTick(
-            final JobRepository<T> repo,
-            final Organization org,
-            final Job<T> job,
-            final Instant scheduled
-    ) {
-        final Optional<Instant> nextScheduled = job.getSchedule().nextRun(Optional.of(scheduled));
-        repo.jobStarted(job.getId(), org, scheduled);
+    private void attemptExecuteAndUpdateRepository(final Instant scheduled) {
+        if (!_cachedJob.isPresent()) {
+            LOGGER.error()
```

Comment:
See note above about `IllegalStateException`; this feels like a similar situation.